client/gutenberg/extensions/slideshow/create-swiper.js

```diff
@@ -25,5 +23,10 @@ export default function createSwiper( container = '.swiper-container', params =
 		setWrapperSize: true,
 		touchStartPreventDefault: false,
 	};
+
```

Comment:
The async loading change creates a regression. `createSwiper()` now returns a `Promise` rather than an instance of `Swiper` which leads to errors when `this.swiperInstance.update()` or`this.swiperInstance.destroy()` are called in `slideshow.js`.

Reply:
Took a shot at resolving the Promise issue in https://github.com/Automattic/wp-calypso/pull/30352/commits/581113b2fd0336d46150169d0694b076311c5f0d. @ockham curious if you had a more elegant approach in mind.

Reply:
Ah, good spot, thanks!

> @ockham curious if you had a more elegant approach in mind.

My Promise-fu isn't the strongest, but how about making the relevant lifecycle methods `async` and `await`ing the result of `this.buildSwiper()`? (According to [cursory](https://stackoverflow.com/questions/47970276/is-using-async-componentdidmount-good) [research](https://www.valentinog.com/blog/how-async-await-in-react/), this seems to be a legit thing to do.) (We don't need to change `buildSwiper` itself since it just passes the promise through).

```diff
diff --git a/client/gutenberg/extensions/slideshow/slideshow.js b/client/gutenberg/extensions/slideshow/slideshow.js
index aef323173f..729b378cd3 100644
--- a/client/gutenberg/extensions/slideshow/slideshow.js
+++ b/client/gutenberg/extensions/slideshow/slideshow.js
@@ -25,11 +25,11 @@ class Slideshow extends Component {
 		this.paginationRef = createRef();
 	}
 
-	componentDidMount() {
-		this.swiperInstance = this.buildSwiper();
+	async componentDidMount() {
+		this.swiperInstance = await this.buildSwiper();
 	}
 
-	componentDidUpdate( prevProps ) {
+	async componentDidUpdate( prevProps ) {
 		const { align, effect, images } = this.props;
 
 		/* A change in alignment or images only needs an update */
@@ -40,7 +40,7 @@ class Slideshow extends Component {
 		if ( effect !== prevProps.effect ) {
 			const { activeIndex } = this.swiperInstance;
 			this.swiperInstance.destroy( true, true );
-			this.swiperInstance = this.buildSwiper( activeIndex );
+			this.swiperInstance = await this.buildSwiper( activeIndex );
 		}
 	}
 
```

(After reverting 581113b2fd0336d46150169d0694b076311c5f0d.)

Reply:
I tried this approach out and it works nicely. Here are a couple of questions for discussion about this proposed approach:

- After reading through https://www.valentinog.com/blog/how-async-await-in-react/, noted that IE support may be a problem: https://caniuse.com/#search=async%20functions. I tried it out in IE and did in fact get `Unhandled promise rejection SyntaxError` in the console.

- Since the Swiper library is fairly large, I'm wondering if there are any negatives to using `await` in the Component's lifecycle methods. If the pause while awaiting the library is noticeable, will this negatively impact the block's UX?

Reply:
> I tried this approach out and it works nicely. Here are a couple of questions for discussion about this proposed approach:
> 
> * After reading through https://www.valentinog.com/blog/how-async-await-in-react/, noted that IE support may be a problem: https://caniuse.com/#search=async%20functions. I tried it out in IE and did in fact get `Unhandled promise rejection SyntaxError` in the console.

Good spot. We use `async`/`await` all over the place in Calypso so I thought we were polyfilling it in the Babel config (which the SDK shares), but I'd forgotten that we're really [importing it in `client/boot/polyfill`](https://github.com/Automattic/wp-calypso/blob/cc03019d364164cbe6a42997b18f22bd471b1650/client/boot/polyfills.js#L5) -- i.e. upon startup.

Ideally, we should be able to use the same ES subset in both Calypso and blocks. We might need to stick to the `.then()` approach in this PR for now so we don't get blocked and work on that separately.

> * Since the Swiper library is fairly large, I'm wondering if there are any negatives to using `await` in the Component's lifecycle methods. If the pause while awaiting the library is noticeable, will this negatively impact the block's UX?

:thinking: Would it make any difference in practice? I'm assuming the `import()`s would only take a while upon first load (and be readily available upon later `createSwiper()` calls).

_tl;dr_ I guess we'll stick with the `.then()`s for now

Reply:
At second glance, we're already using `async`/`await` in a few other spots in `client/gutenberg/extensions/`. I wonder if that means they're all broken in IE :slightly_frowning_face:

Reply:
Anyway, let's stick with what we have right now. This is also material for a separate discussion. I'll file an issue.

Reply:
#30628